R/tree-map-alignment.R

```diff
@@ -455,16 +460,344 @@ find_best_shift_icp = function(pred, obs) {
+  # Returns:
+  #   from_idx2_to_closest_idx1 = N2-by-1 vector containing the indices of the closest
+  #           object in dataset 1 for each object in dataset 2.
+  #   from_idx2_to_min_dist1 = N2-by-1 vector containing the distances to the closest
+  #           object in dataset 1 for each object in dataset 2.
+  #   n_of_matches = number of matching object pairs based on r_thres.
+  xy_mat1 = as.matrix(xy_mat1)
+  xy_mat2 = as.matrix(xy_mat2)
+  colnames(xy_mat2) = NULL
+  xy_mat1_t = t(xy_mat1)
+  rotated_xy1 = t(R_mat %*% xy_mat1_t)
+  t_vect = as.vector(as.numeric(t_vect))
+  xy_mat1_transformed = t(t(rotated_xy1) + t_vect) # N-by-2
+  # Pair-wise locations between each object in dataset 1 and 2. Each row is
+  # one object in dataset 2 and each column in one object in dataset 1.
+  transformed_loc_p_dist = cdist(xy_mat2, xy_mat1_transformed)
+  # Finding the closest object in transformed point cloud 1 for each object
+  # in point cloud 2
+  from_idx2_to_closest_idx1 = apply(transformed_loc_p_dist, 1, which.min)
+  from_idx2_to_min_dist1 = transformed_loc_p_dist[
+    cbind(1:nrow(transformed_loc_p_dist), from_idx2_to_closest_idx1)
+  ]
+
+  # Computing the number of matches, i.e., object pairs whose distance is
+  # below the set criterion
+  less_than_thresh = from_idx2_to_min_dist1 < r_thres
+  n_of_matches = sum(less_than_thresh)
+  return(list(from_idx2_to_closest_idx1, from_idx2_to_min_dist1, n_of_matches))
+}
 
+
+# Implementation of the following MATLAB function
+# https://gitlab.com/fgi_nls/public/2d-registration/-/blob/main/fit_euclidean_transformation.m?ref_type=heads
+find_best_shift_hyyppa = function(pred, obs, R_local = 10, k = 20, r_thresh = 1.0, max_iters = 200) {
+  # Rename the variables for consistency with the reference code and remove unneeded columns
+  xy_mat1 = pred[c("x", "y")]
+  xy_mat2 = obs[c("x", "y")]
+
+  # Centering the coordinates before matching to improve numerical stability
+  xy_mat1_mean = colMeans(xy_mat1[c("x", "y")])
+  xy_mat2_mean = colMeans(xy_mat2[c("x", "y")])
+  # The double transpose seems to be the easiest way to broadcast subtraction
+  # given the recycling rules
+  xy_mat1 = t(t(xy_mat1) - xy_mat1_mean)
+  xy_mat2 = t(t(xy_mat2) - xy_mat2_mean)
+
+  # Number of objects detected from each point cloud
+  N_objects_vect = c(nrow(xy_mat1), nrow(xy_mat2))
+
+  if (N_objects_vect[2] < k) {
+    k = N_objects_vect[2]
+    msg = paste0(
+      "Number of objects in dataset 2 (and thus the potential number of matches) ",
+      "is smaller than given parameter k. k was set to ",
+      N_objects_vect[2]
+    )
+    # TODO make this a warning
+    print(msg)
+  }
+
+  # First, we construct the feature descriptor for each object in each of
+  # the point clouds.
+  feature_info1 = compute_feature_descriptors(xy_mat1, R_local)
+  feature_info2 = compute_feature_descriptors(xy_mat2, R_local)
+  # Extract out the components of the result
+  feat_desc_mat1 = feature_info1[[1]]
```

Comment:
Here you could reference the list elements by name as opposed to index, if you named them when the list is created by `compute_feature_descriptors`.